segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,413 @@
+        PosixFileAttributes attrs = Files.readAttributes(Paths.get(config.getFilesystemRoot(), streamSegmentName),
+                PosixFileAttributes.class);
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName, attrs.size(),
+                !(attrs.permissions().contains(OWNER_WRITE)), false,
+                new ImmutableDate(attrs.creationTime().toMillis()));
+
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        return Files.exists(Paths.get(config.getFilesystemRoot(), streamSegmentName));
+    }
+
+    @SneakyThrows
+    private SegmentProperties syncCreate(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+        log.info("Creating Segment {}", streamSegmentName);
+        Set<PosixFilePermission> perms = new HashSet<>();
+        // add permission as rw-r--r-- 644
+        perms.add(PosixFilePermission.OWNER_WRITE);
+        perms.add(PosixFilePermission.OWNER_READ);
+        perms.add(PosixFilePermission.GROUP_READ);
+        perms.add(PosixFilePermission.OTHERS_READ);
+        FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        Files.createDirectories(path.getParent());
+        Files.createFile(path, fileAttributes);
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return this.syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    @SneakyThrows
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        if (handle.isReadOnly()) {
+            throw new IllegalArgumentException("Write called on a readonly handle of segment "
+                    + handle.getSegmentName());
+        }
+
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+        //Fix for the case where Pravega runs as super user privileges.
+        // This means that writes to readonly files also succeed. We need to explicitly check permissions in this case.
+        if ( !isWritableFile(path)) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        long fileSize = path.toFile().length();
+        if (fileSize < offset) {
+            throw new BadOffsetException(handle.getSegmentName(), fileSize, offset);
+        } else {
+            try (FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE);
+                 ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                while (length != 0) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
```

Comment:
Like I said a few times before, you can end up in an infinite loop in here if you start getting 0 for `bytesWritten`